fms/models/gpt_oss.py

```diff
@@ -0,0 +1,637 @@
+        ):
+            self.rot_emb.device = device
+
+    def forward(
+        self,
+        x_in,
+        position_ids=None,
+        past_key_value_states=None,
+        use_cache=False,
+        **attn_kwargs: Unpack[AttentionKwargs],
+    ):
+        # Embed the given vocabulary indices using the given attention mask, with pre-/post-norm and dropout as specified
+        # x_in: batch_size x seq_len x emb_dim if input is already embedded, otherwise batch_size x seq_len
+        # mask: batch_size x seq_len x seq_len
+        # bias: nheads x seq_len x seq_len
+        attn_kwargs.update({"attn_name": "sdpa_with_sinks"})
+
+        if past_key_value_states is None or len(past_key_value_states) == 0:
+            past_key_value_states = [None for _ in range(len(self.layers))]
+
+        if x_in.dim() == 2:  # input is not already embedded
+            x_in = self.embedding(x_in)
+
+        # this is the output cache for all the decoder layers
+        present_key_value_states = []
+
+        for i, layer in enumerate(self.layers):
+            if layer.config.sliding_window > 0:  # type: ignore
+                attn_kwargs.update({"sliding_window": layer.config.sliding_window})  # type: ignore
+            else:
+                attn_kwargs.update({"sliding_window": 0})
+            output = layer(
+                x=x_in,
+                position_ids=position_ids,
+                past_key_value_state=past_key_value_states[i],
+                use_cache=use_cache,
+                **attn_kwargs,
+            )
+
+            if use_cache:
+                x_in, present_key_value_state = output
+                present_key_value_states.append(present_key_value_state)
+
+            else:
+                x_in = output
+
+        dec_out = x_in
+        dec_out = self.dec_norm(dec_out)
+        if self.config.p_dropout:
+            dec_out = self.dropout(dec_out)
+
+        return dec_out, present_key_value_states
+
+
+class GptOss(nn.Module):
+    def __init__(
+        self,
+        config: Optional[GptOssConfig] = None,
+        distributed_strategy: DistributedStrategy = NoOpStrategy,
+        **kwargs,
+    ):
+        super(GptOss, self).__init__()
+        if config is not None:
+            self.config = config
+        else:
+            self.config = GptOssConfig()
+        self.config = self.config.updated(**kwargs)
+        self.distributed_strategy = distributed_strategy
+
+        self.base_model = GptOssHeadless(self.config, self.distributed_strategy)
+        self.head = nn.Linear(
+            self.config.emb_dim, self.config.src_vocab_size, bias=False
+        )
+
+        # handle assignment of non-meta weights to meta parameters
+        if self.head.weight.device == torch.device("meta"):
+            assert self.base_model.embedding.weight != torch.device("meta")
+            self.head.weight = self.base_model.embedding.weight
+        else:
+            self.base_model.embedding.weight = self.head.weight
```

Comment:
this is only needed if we're tying weights between embeddings and head, are we doing that in this model?